src/cunumeric/fft/fft.cu

```diff
@@ -0,0 +1,392 @@
+ * limitations under the License.
+ *
+ */
+#include <csignal>
+
+#include "cunumeric/fft/fft.h"
+#include "cunumeric/fft/fft_template.inl"
+
+#include "cunumeric/cuda_help.h"
+
+namespace cunumeric {
+
+using namespace Legion;
+using namespace legate;
+
+using dim_t = long long int;
+
+template <int DIM>
+__host__ static inline void cufft_operation(void* output,
+                                            void* input,
+                                            const Rect<DIM>& out_rect,
+                                            const Rect<DIM>& in_rect,
+                                            std::vector<int64_t>& axes,
+                                            fftType type,
+                                            fftDirection direction)
+{
+  auto stream = get_cached_stream();
+
+  size_t workarea_size = 0;
+  size_t num_elements;
+  dim_t n[DIM];
+  dim_t inembed[DIM];
+  dim_t onembed[DIM];
+
+  const Point<DIM> zero   = Point<DIM>::ZEROES();
+  const Point<DIM> one    = Point<DIM>::ONES();
+  Point<DIM> fft_size_in  = in_rect.hi - in_rect.lo + one;
+  Point<DIM> fft_size_out = out_rect.hi - out_rect.lo + one;
+  num_elements            = 1;
+  for (int i = 0; i < DIM; ++i) {
+    n[i] =
+      (type == fftType::FFT_R2C || type == fftType::FFT_D2Z) ? fft_size_in[i] : fft_size_out[i];
+    inembed[i] = fft_size_in[i];
+    onembed[i] = fft_size_out[i];
+    num_elements *= n[i];
+  }
+
+  // Create the plan
+  cufftHandle plan;
+  CHECK_CUFFT(cufftCreate(&plan));
+  CHECK_CUFFT(cufftSetAutoAllocation(plan, 0 /*we'll do the allocation*/));
+  CHECK_CUFFT(cufftSetStream(plan, stream));
+
+  // Create the plan and allocate a temporary buffer for it if it needs one
+  CHECK_CUFFT(cufftMakePlanMany64(
+    plan, DIM, n, inembed, 1, 1, onembed, 1, 1, (cufftType)type, 1, &workarea_size));
+
+  DeferredBuffer<uint8_t, 1> workarea_buffer;
+  if (workarea_size > 0) {
+    const Point<1> zero1d(0);
+    workarea_buffer = DeferredBuffer<uint8_t, 1>(Rect<1>(zero1d, Point<1>(workarea_size - 1)),
+                                                 Memory::GPU_FB_MEM,
+                                                 nullptr /*initial*/,
+                                                 128 /*alignment*/);
+    void* workarea  = workarea_buffer.ptr(zero1d);
+    CHECK_CUFFT(cufftSetWorkArea(plan, workarea));
+  }
+
+  // FFT the input data
+  CHECK_CUFFT(cufftXtExec(plan, input, output, (int)direction));
+
+  // Clean up our resources, DeferredBuffers are cleaned up by Legion
+  CHECK_CUFFT(cufftDestroy(plan));
+}
+
+template <int DIM, typename OUTPUT_TYPE, typename INPUT_TYPE>
+struct cufft_axes_plan {
+  __host__ static inline void execute(cufftHandle plan,
+                                      // AccessorWO<OUTPUT_TYPE, DIM> out,
+                                      // AccessorRO<INPUT_TYPE, DIM> in,
+                                      DeferredBuffer<OUTPUT_TYPE, DIM>& out,
```

Comment:
Legion's `DeferredBuffer` is re-exported by Legate on the name of `legate::Buffer`. So you can just type `Buffer<OUTPUT_TYPE, DIM>`.